lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,199 @@
+                                           InstsToErase, Builder);
+      }
+      // Handle load instruction which is following the bitcast in the pattern
+      else if (auto *LI = dyn_cast<LoadInst>(U)) {
+        Value *LocalValue = Builder->CreateLoad(NewVecTy, NewInst);
+        Changed |= lowerBitCastToNonStdVec(LI, LocalValue, OldVecTy,
+                                           InstsToErase, Builder);
+      }
+      // Handle extractelement instruction which is following the load
+      else if (auto *EEI = dyn_cast<ExtractElementInst>(U)) {
+        uint64_t NumElemsInOldVec = OldVecTy->getElementCount().getValue();
+        uint64_t NumElemsInNewVec = NewVecTy->getElementCount().getValue();
+        uint64_t ElemIdx =
+            cast<ConstantInt>(EEI->getIndexOperand())->getSExtValue() /
+            (NumElemsInOldVec / NumElemsInNewVec);
+        Value *LocalValue = Builder->CreateExtractElement(NewInst, ElemIdx);
+        LocalValue =
+            Builder->CreateTrunc(LocalValue, OldVecTy->getElementType());
+        Changed |= lowerBitCastToNonStdVec(EEI, LocalValue, OldVecTy,
+                                           InstsToErase, Builder);
+      }
+    }
+  }
+  InstsToErase.push_back(OldInst);
+  if (!Changed)
+    OldInst->replaceAllUsesWith(NewInst);
+  return true;
+}
+
+class SPIRVLowerBitCastToNonStandardTypePass
+    : public llvm::PassInfoMixin<SPIRVLowerBitCastToNonStandardTypePass> {
+public:
+  SPIRVLowerBitCastToNonStandardTypePass() {}
+
+  PreservedAnalyses
+  runLowerBitCastToNonStandardType(Function &F, FunctionAnalysisManager &FAM) {
+    // This pass doesn't cover all possible uses of non-standard types, only
+    // known. We assume that bad type won't be passed to a function as
+    // parameter, since it added by an optimization.
+    bool Changed = false;
+
+    std::vector<Instruction *> BCastsToNonStdVec;
+    std::vector<Instruction *> InstsToErase;
+    for (auto &BB : F)
+      for (auto &I : BB) {
+        auto *BC = dyn_cast<BitCastInst>(&I);
+        if (!BC)
+          continue;
+        VectorType *SrcVecTy = getVectorType(BC->getSrcTy());
+        if (SrcVecTy) {
+          uint64_t NumElemsInSrcVec = SrcVecTy->getElementCount().getValue();
+          if (!isValidVectorSize(NumElemsInSrcVec))
+            report_fatal_error("Unsupported vector type with the size of: " +
+                                   std::to_string(NumElemsInSrcVec),
+                               false);
+        }
+        VectorType *DestVecTy = getVectorType(BC->getDestTy());
+        if (DestVecTy) {
+          uint64_t NumElemsInDestVec = DestVecTy->getElementCount().getValue();
+          if (!isValidVectorSize(NumElemsInDestVec))
+            BCastsToNonStdVec.push_back(&I);
+        }
+      }
+    IRBuilder<> Builder(F.getContext());
+    for (auto &I : BCastsToNonStdVec) {
+      Value *NewValue = I->getOperand(0);
+      VectorType *OldVecTy = getVectorType(I->getType());
+      Builder.SetInsertPoint(I);
```

Comment:
Could you please move this call into the beginning of lowerBitCastToNonStdVec? Dependent instructions are not guarantied to be consecutive in IR.